*: add changelog for 0.99.5 (#2787)
nspcc-dev/neo-go: CHANGELOG.md

This document outlines major changes between releases.

## 0.99.5 "Underestimation" (11 Nov 2022)

> It wasn't long since the previous release before a juicy set of features and

Reviewer: ```suggestion
It wasn't long since the previous release until a juicy set of features and
```

> optimisations came on board, so we've decided to change our release plan and

Reviewer: ```suggestion
optimisations came on board. So, we've decided to change our release plan and
```

supply one more 3.4.0 compatible version. It comes up with a massive suite
of network P2P and broadcast improvements. We've carefully investigated message
sending timeouts on benchmarks, reworked the broadcast logic and significantly

> reduced delays in message sending and amount of useless traffic between our

Reviewer: ```suggestion
reduced delays in message sending and the amount of useless traffic between our
```

nodes. For smart-contract developers and RPC client users several new features
were added. These are automatic RPC bindings generator, poll-based and

> websocket-based transactions awaiting and extended web-socket client

Reviewer: ```suggestion
websocket-based transactions awaiting and an extended web-socket client
```

subscriptions interface. For better debugging experience we've improved state
related functionality of VM CLI and added an ability to reset chain state to a
particular height.

An important dBFT initialization bug that caused timestamp-related errors on
block addition was fixed. Also, those who use NeoGo node with

> EnableCORSWorkaround setting may want to update to be able to handle web-socket

Reviewer: ```suggestion
EnableCORSWorkaround setting may want an update to be able to handle web-socket
```

Author: Let it be the `may want to update the node`, is it OK?

> requests with Origin set in header. Finally, we're kindly asking our users to

Reviewer: ```suggestion
requests with Origin set in the header. Finally, we are kindly asking our users to
```

> make acquaintance with our roadmap file where deprecated code removal schedule

Reviewer: ```suggestion
have a look at our roadmap file where deprecated code removal schedule
```

> is placed.

Reviewer: ```suggestion
is attached.
```

This version is fully compatible with C# node 3.4.0 and does not require
resynchronization on upgrade. This time for sure, the next release is planned
to be compatible with 3.5.0.

New features:
 * logarithmic gossip fan-out, new `BroadcastFactor` application setting and
   adaptive peer requests are added to optimise broadcasting process (#608,
   #2678, #2743)
 * Prometheus histograms with handling time are added for P2P commands (#2750)
   and RPC calls (#2784)
 * transaction awaiting is added to RPC clients (#2740, #2749)
 * RPC bindings generator (#2705) is added for both safe (#2766) and
   state-changing (#2769, #2778) methods with initial iterator support (#2783)
 * the ability to reset chain state to a particular height is added (#2576,
   #2773)

Behaviour changes:
 * network packets sending is now limited in time (#2738)
 * broadcast messages will be sent concurrently to each peer (#2741)
 * ping messages are broadcast now (#2742)
 * transactions requested by dBFT will be accepted on demand only (#2746, #2748,

>    nspcc-dev/dbft#63)

Reviewer 2: As well as everything above, it's not something end user can immediately feel or need to take care about.

 * Aspidochelone fork block for NeoFS sidechain mainnet configuration is
   rescheduled (#2754, #2776)

>  * `NOT` opcode for Bool emission is now used to reduce resulting script

Reviewer 2: While this is not so much, just an optimization/improvement.

   cost (#2762)

>  * RPC server will now handle any Origin in websocket connection when

Reviewer 2: This is a behavior change.

   EnableCORSWorkaround setting is on (#2772, #2724)

Improvements:
 * network code refactoring (#2742)
 * VM CLI's state-based functionality improvements (extended signers and
   arguments parsing, enabled logging) and additional commands (`loadtx`,
   `loaddeployed`, `jump`) (#2729, #2606, #2740)
 * incoming P2P transactions are now processed concurrently and buffer for
   inventory hashes is reusable (#2755)
 * `GetData` and `GetBlocksByIndex` replies are now batched on a TCP-level
   (#2757)

>  * optimisation of incoming P2P transactions and blocks filtering went in

Reviewer: ```suggestion
 * optimisation of incoming P2P transactions and blocks filtering introduced in
```
if I get it right?

Author: Just `are introduced` to keep the common style.

   (#2758, #2759)
 * batch size of broadcast transactions hashes is adjusted (#2760)
 * improve peer reconnections logic and send `GetAddr` messages more often
   (#2761, #2745)
 * clarify documentation for wallet configuration file (#2763)
 * WS RPC client subscriptions API is extended to accept custom subscriber
   channels (#2756, #2764)
 * Koblitz curve implementation of btcd/btcec was replaced with decred/secp256k1
   due to GitHub security warning (#2781)

>  * deprecated code removal schedule was added to a roadmap (#2785)

Reviewer 2: Not directly something we release, it's OK to only mention this in the introduction.

This document outlines major changes between releases.

## 0.99.5 "Underestimation" (11 Nov 2022)

> A long time has passed since the previous release, a juicy set of features and

Reviewer 2: A month. Not very long, I'd say.

reduced delays in message sending and amount of useless traffic between our
nodes. For smart-contract developers and RPC client users several new features
were added. These are automatic RPC bindings generator, poll-based and
websocket-based transactions awaiting and extended web-socket client
subscriptions interface. For better debugging experience we've improved state
related functionality of VM CLI and added an ability to reset chain state to a
particular height.

An important dBFT initialization bug that caused timestamp-related errors on
block addition was fixed. Also, those who use NeoGo node with
EnableCORSWorkaround setting may want to update to be able to handle web-socket
requests with Origin set in header. Finally, we're kindly asking our users to
make acquaintance with our roadmap file where deprecated code removal schedule
is placed.

This version is fully compatible with C# node 3.4.0 and does not require
resynchronization on upgrade. This time for sure, the next release is planned
to be compatible with 3.5.0.

New features:
 * logarithmic gossip fan-out, new `BroadcastFactor` application setting and

>    adaptive peer requests are added to optimise broadcasting process (#608,

Reviewer 2: Usually I'm using PR numbers only, but issues are OK too.

Author: I didn't include all issues, but some of them contain useful information or duscussions, thus I've included thim. Although I'm not sure about it, maybe we really need PRs only to be present here.

Reviewer 2: It's not a strict policy, both can be referenced if needed.

   #2678, #2743)
 * Prometheus histograms with handling time are added for P2P commands (#2750)
   and RPC calls (#2784)
 * transaction awaiting is added to RPC clients (#2740, #2749)
 * RPC bindings generator (#2705) is added for both safe (#2766) and
   state-changing (#2769, #2778) methods with initial iterator support (#2783)
 * the ability to reset chain state to a particular height is added (#2576,
   #2773)

Behaviour changes:
 * network packets sending is now limited in time (#2738)
 * broadcast messages will be sent concurrently to each peer (#2741)

>  * ping messages are broadcast now (#2742)

Reviewer 2: Not sure it's a correct description.

Author: But don't we use a broadcast logic for them since https://github.com/nspcc-dev/neo-go/pull/2742/commits/b345581c72e15ef151555b30e94bbd4735daa3d2?

Reviewer 2: Well, they were broadcasted, just to a bigger set of recipients. Now this set is somewhat smaller.

Author: Maybe `ping messages are broadcast using the common broadcast logic` will suite?

Reviewer 2: ok

Author: Fixed.

>  * transactions requested by dBFT will be accepted on demand only (#2746, #2748,

Reviewer 2: Random unrelated thought: we have a list of requested transactions in the network server and we could only push requested ones to dbft.

Author: Not sure, should I rephrase it?

Reviewer 2: #2788.

   nspcc-dev/dbft#63)
 * Aspidochelone fork block for NeoFS sidechain mainnet configuration is
   rescheduled (#2754, #2776)
 * `NOT` opcode for Bool emission is now used to reduce resulting script
   cost (#2762)
 * RPC server will now handle any Origin in websocket connection when
   EnableCORSWorkaround setting is on (#2772, #2724)

Improvements:
 * network code refactoring (#2742)
 * VM CLI's state-based functionality improvements (extended signers and arguments
   parsing, enabled logging) and additional commands (`loadtx`, `loaddeployed`,
   `jump`) (#2729, #2606, #2740)
 * incoming P2P transactions are now processed concurrently and buffer for
   inventory hashes is reusable (#2755)

>  * investigation of the network I/O timeouts during bench test was made (#2744) // TODO: do we need to put this issue here? We have all the related PRs mentioned already.

Reviewer 2: Usually  I only pick merged code changes. We care about our deliverable changes here, it's all code.

 * `GetData` and `GetBlocksByIndex` replies are now batched on a TCP-level (#2757)

>  * incoming P2P transactions are checked against currently processing ones (#2758)

Reviewer 2: Too fine-grained to me.

 * incoming P2P blocks check optimisation for recently added blocks is added (#2759)
 * batch size of broadcast transactions hashes is adjusted (#2760)
 * improve peer reconnections logic and send `GetAddr` messages more often
   (#2761, #2745)
 * clarify documentation for wallet configuration file (#2763)
 * WS RPC client subscriptions API is extended to accept custom subscriber
   channels (#2756, #2764)

>  * security vulnerability of Koblitz curve implementation was fixed with dropped

Reviewer 2: I think there was no vulnerability in our case.

Author: You're right, it sounds like we had it, so I've rephrased the sentence.
